src/ray/id.h

```diff
@@ -2,44 +2,128 @@
+  static T from_random();
+  static T from_binary(const std::string &binary);
+  static const T &nil();
+  static size_t size() { return T::size(); }
+
   size_t hash() const;
   bool is_nil() const;
-  bool operator==(const UniqueID &rhs) const;
-  bool operator!=(const UniqueID &rhs) const;
+  bool operator==(const BaseID &rhs) const;
+  bool operator!=(const BaseID &rhs) const;
   const uint8_t *data() const;
-  static size_t size();
   std::string binary() const;
   std::string hex() const;
-  plasma::UniqueID to_plasma_id() const;
 
- private:
+ protected:
+  BaseID(const std::string &binary) {
+    std::memcpy(const_cast<uint8_t *>(this->data()), binary.data(), T::size());
+  }
+  // All IDs are immutable for hash evaluations. mutable_data is only allow to use
+  // in construction time, so this function is protected.
+  uint8_t *mutable_data();
+  // For lazy evaluation, be careful to have one Id contained in another.
+  // This hash code will be duplicated.
+  mutable size_t hash_ = 0;
+};
+
+class UniqueID : public BaseID<UniqueID> {
+ public:
+  UniqueID() : BaseID(){};
+  static size_t size() { return kUniqueIDSize; }
+
+ protected:
   UniqueID(const std::string &binary);
 
  protected:
   uint8_t id_[kUniqueIDSize];
-  mutable size_t hash_ = 0;
 };
 
-static_assert(std::is_standard_layout<UniqueID>::value, "UniqueID must be standard");
+class TaskID : public BaseID<TaskID> {
+ public:
+  TaskID() : BaseID() {}
+  static size_t size() { return kTaskIDSize; }
+  static TaskID GetDriverTaskID(const DriverID &driver_id);
+
+ private:
+  uint8_t id_[kTaskIDSize];
+};
+
+class ObjectID : public BaseID<ObjectID> {
+ public:
+  ObjectID() : BaseID() {}
+  static size_t size() { return kUniqueIDSize; }
+  plasma::ObjectID to_plasma_id() const;
+  ObjectID(const plasma::UniqueID &from);
+
+  /// Get the index of this object in the task that created it.
+  ///
+  /// \return The index of object creation according to the task that created
+  /// this object. This is positive if the task returned the object and negative
+  /// if created by a put.
+  int32_t object_index() const { return index_; }
+
+  /// Compute the task ID of the task that created the object.
+  ///
+  /// \return The task ID of the task that created this object.
+  TaskID task_id() const;
+
+  /// Compute the object ID of an object put by the task.
+  ///
+  /// \param task_id The task ID of the task that created the object.
+  /// \param index What index of the object put in the task.
+  /// \return The computed object ID.
+  static ObjectID for_put(const TaskID &task_id, int64_t put_index);
+
+  /// Compute the object ID of an object returned by the task.
+  ///
+  /// \param task_id The task ID of the task that created the object.
+  /// \param return_index What index of the object returned by in the task.
+  /// \return The computed object ID.
+  static ObjectID for_task_return(const TaskID &task_id, int64_t return_index);
+
+ private:
+  uint8_t id_[kTaskIDSize];
```

Comment:
could we use TaskID here instead of raw byte array?
or if it is indeed necessary, maybe rename id_ to task_id_?

Reply:
@zhijunfu In the first version, yes, it is a TaskID. However, if this is the case, TaskID cannot have cached hash value, otherwise the structure will be very strange, e.g. There are 2 hash code cache at the beginning of ObjectID. This may gone, we the ID schema is fixed. We can remove the hash cache code.